src/main/java/org/cloudfoundry/promregator/cfaccessor/ReactiveCFAccessorImpl.java

```diff
@@ -150,7 +146,7 @@ private DefaultConnectionContext connectionContext(ProxyConfiguration proxyConfi
 	}
 
 	private PasswordGrantTokenProvider tokenProvider() {
-		return PasswordGrantTokenProvider.builder().password(this.password).username(this.username).build();
+		return PasswordGrantTokenProvider.builder().password(this.cfApiCredentials.getPassword()).username(this.cfApiCredentials.getUsername()).build();
```

Comment:
Looking at this now, it comes to my mind that you are using a feature which we haven't built intentionally, though :) 
BTW: How you are detecting that the password expired? Just by the fact that you can't connect to CF API anymore? Would there be alternative options, e.g. event mechanism?

Reply:
Unintentional features are sometime the _very best_ kind. The thought here was to leave it to the implementer of extension based on the operation of their credentials mechanism:

- always pull - get new creds every time before calling cf api
- test-before-use - ensure that cached creds are valid and pull new ones if necessary before calling cf api
- pull on failure - assume cached creds are valid and pull new ones if a 401/403 (or whatever custom response) is received from cf api
- message on change - receive notification from platform/service which causes cached credentials to be updated before calling cf api
- any number of other custom solutions which cause cached creds to be updated

Reply:
Well, yes, but this decision is only possible, if the library is asking the TokenProvider in first place. I would assume that `cf-java-client` is caching the JWT, so there are only very occasional inquiries coming to the token provider (hopefully!).
I do not want to intervene in the way the implementation is doing things here - they should have their full freedom how to handle caching. It's more the fear that due to several authentication attempts, `cf-java-client` might be asking the extension *too late*.

Reply:
A lot of time has passed after my last comment. The reply to my doubts are is simple: As long as the JWT is valid (i.e. "not expired"), it does not matter that the password might have changed: The JWT is still valid and no username/password is necessary anymore. 
The only thing which is important is that upon each request of a new JWT, the `cf-java-client` library needs to query the registered `PasswordGrantTokenProvider` - which is what it does.